src/modules/form/components/DynamicFormField.tsx

```diff
@@ -0,0 +1,276 @@
+import { isEmpty, pick } from 'lodash-es';
+import React, { cloneElement, ReactNode, useCallback } from 'react';
+
+import {
+  FormDefinitionHandlers,
+  getSafeLinkId,
+} from '../hooks/useFormDefinitionHandlers';
+import {
+  ChangeType,
+  ItemChangedFn,
+  OverrideableDynamicFieldProps,
+  PickListArgs,
+  SeveralItemsChangedFn,
+} from '../types';
+import {
+  buildCommonInputProps,
+  getAllChildLinkIds,
+  transformSubmitValues,
+} from '../util/formUtil';
+
+import AutofillFormItemWrapper from './AutofillFormItemWrapper';
+import DependentFormItemWrapper from './DependentFormItemWrapper';
+import DynamicField from './DynamicField';
+import DynamicGroup from './DynamicGroup';
+import ValueWrapper from './ValueWrapper';
+import { formatCurrency } from '@/modules/hmis/hmisUtil';
+import {
+  Component,
+  FormItem,
+  ItemType,
+  ServiceDetailType,
+} from '@/types/gqlTypes';
+
+export const renderItemWithWrappers = (
+  renderChild: (disabled?: boolean) => ReactNode,
+  item: FormItem,
+  handlers: FormDefinitionHandlers
+) => {
+  const { disabledDependencyMap, autofillInvertedDependencyMap } = handlers;
+
+  if (item.hidden) return null;
+
+  const hasDependencies =
+    disabledDependencyMap[item.linkId] ||
+    !isEmpty(item.enableWhen) ||
+    item.item?.every((item) => item.enableWhen);
+  const hasAutofill =
+    autofillInvertedDependencyMap[item.linkId] || !isEmpty(item.autofillValues);
+
+  if (hasDependencies && hasAutofill) {
+    return (
+      <DependentFormItemWrapper handlers={handlers} item={item}>
+        {(disabled) => (
+          <AutofillFormItemWrapper handlers={handlers} item={item}>
+            {() => renderChild(disabled)}
+          </AutofillFormItemWrapper>
+        )}
+      </DependentFormItemWrapper>
+    );
+  }
+
+  if (hasDependencies) {
+    return (
+      <DependentFormItemWrapper handlers={handlers} item={item}>
+        {renderChild}
+      </DependentFormItemWrapper>
+    );
+  }
+
+  if (hasAutofill) {
+    return (
+      <AutofillFormItemWrapper handlers={handlers} item={item}>
+        {() => renderChild()}
+      </AutofillFormItemWrapper>
+    );
+  }
+
+  return renderChild();
+};
+
+export interface Props {
+  handlers: FormDefinitionHandlers;
+  clientId?: string;
+  horizontal?: boolean;
+  warnIfEmpty?: boolean;
+  locked?: boolean;
+  bulk?: boolean;
+  visible?: boolean;
+  pickListArgs?: PickListArgs;
+  nestingLevel: number;
+  item: FormItem;
+  props?: OverrideableDynamicFieldProps;
+  renderFn?: (children: ReactNode) => ReactNode;
+}
+
+const DynamicFormField: React.FC<Props> = ({
```

Comment:
My understanding is the like the stack here is something like this. Does that seem right?
```js
DynamicFormField(renderFn)(
  // renderItemWithWrappers
  DependentFormItemWrapper(
    AutofillFormItemWrapper(
      //itemGroup, table etc
      (item) => item.children.map (
        ValueWrapper(DynamicField)
      )
    )
  )
)
```

Reply:
Thanks for the write up! The description is helpful to get oriented in the form mechanics. Maybe the above could become inline comments/docs? (if it's not already there)